libbeat/common/wait/timer.go

```diff
@@ -0,0 +1,143 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package wait
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common/atomic"
+)
+
+// Timer represents a timer implementation.
+type Timer interface {
+	Reset(time.Duration) bool
+	Start()
+	Stop() bool
+	Wait() <-chan time.Time
+}
+
+// Strategy is the strategy to be used to find out the time to wait before making a call.
+type Strategy func() time.Duration
+
+// MinWaitAndJitter takes a minimal wait time and jitter and will return a duration that will be at
+// least equal of bigger than the initial wait time.
+func MinWaitAndJitter(d, jitter time.Duration) Strategy {
+	return func() time.Duration {
+		return d + time.Duration(rand.Int63n(int64(jitter)))
+	}
+}
+
+// RandomDelay takes a maximum duration and will return a random values which will be
+// max <= values > 0.
+func RandomDelay(max time.Duration) Strategy {
+	return func() time.Duration {
+		return time.Duration(int64(max))
+	}
+}
+
+// Const waits for a fixed duration.
+func Const(d time.Duration) Strategy {
+	return func() time.Duration {
+		return d
+	}
+}
+
+// PeriodicTimer uses a two Waiter strategy, the first will decide how much time to wait before
+// receive the initial tick and the other will be how much time to wait before doing the other calls,
+// this can be used to introduce more randomness in the frequency of the calls for outside system.
+// This can be useful if you want to better distribute calls that could affect the performance of
+// other system.
+type PeriodicTimer struct {
+	c        chan time.Time
+	done     chan struct{}
+	initial  Strategy
+	periodic Strategy
+	period   time.Duration
+	running  atomic.Bool
+}
+
+// NewPeriodicTimer returns a wait, allowing to wait for a minimum time and a random amount.
+func NewPeriodicTimer(initial, periodic Strategy) *PeriodicTimer {
+	jt := &PeriodicTimer{
+		c:        make(chan time.Time),
+		period:   initial(),
+		periodic: periodic,
+		running:  atomic.MakeBool(false),
+	}
+	return jt
+}
+
+// Start starts the timer.
+// NOTE: Starts should not be called at high frequency.
+func (jt *PeriodicTimer) Start() {
+	if !jt.running.Load() {
+		jt.running.Store(true)
+		jt.done = make(chan struct{})
+		jt.c = make(chan time.Time)
+
+		go func() {
+			jt.running.Store(false)
+			jt.startTimer(jt.period)
+		}()
+	}
+}
+
+func (jt *PeriodicTimer) startTimer(period time.Duration) {
```

Comment:
name 'start' indicates to me that a go-routine is started. Maybe call the timerLoop?